cpp/csp/cppnodes/statsimpl.h

```diff
@@ -1437,30 +1444,39 @@ class Rank
                 }
             }
         #else
-            if( likely( !isnan( m_lastval ) && m_tree.size() > 0 ) )
+            if( likely( !isnan( m_lastval ) && ( ( m_method == MAX && m_maxtree.size() > 0 ) || m_mintree.size() > 0 ) ) )
             {
                 switch( m_method )
                 {
                     case MIN:
                     {
-                        return std::distance( m_tree.begin(), m_tree.find( m_lastval ) );
+                        if ( m_mintree.size() == 1 )
+                            return 0;
+                        return m_mintree.get<0>().find_rank( m_lastval );
                     }
                     case MAX:
                     {
-                        auto end_range = m_tree.equal_range( m_lastval ).second;
-                        return std::distance( m_tree.begin(), std::prev( end_range ) );
+                        if ( m_maxtree.size() == 1 )
+                            return 0;
+                        return m_maxtree.size() - 1 - m_maxtree.get<0>().find_rank( m_lastval );
                     }
                     case AVG:
                     {
-                        auto range = m_tree.equal_range( m_lastval );
-                        return std::distance( m_tree.begin(), range.first ) + ( double )std::distance( range.first, std::prev( range.second ) ) / 2;
+                        if ( m_mintree.size() == 1 )
+                            return 0;
+                        
+                        int min_rank = m_mintree.get<0>().find_rank( m_lastval );
+                        int max_rank = min_rank;
+                        auto it = m_mintree.get<0>().nth( min_rank );
+                        it++;
+                        for( ; it != m_mintree.end() && *it == m_lastval ; it++ ) max_rank++;
```

Comment:
Iterating on the range is _in theory_ O(n) in the worst-case if you have n equal elements. We could use a combination of `equal_range` and `std::distance` between the boost iterators to make this O(log n). However, in the general case, the iteration is going to be way faster as it avoids 2 function calls and there are likely no duplicate values or very few.

So I prefer the current approach, but I'd make a comment explaining that even though this loop looks costly, it's almost always only executed once.